packages/spindle-ui/src/SegmentedControl/SegmentedControl.tsx

```diff
@@ -0,0 +1,192 @@
+import React, {
+  RefObject,
+  createRef,
+  useCallback,
+  useEffect,
+  useLayoutEffect,
+  useRef,
+  useState,
+} from 'react';
+
+type Option = {
+  id: string;
+  label: string;
+};
+
+type Size = 'medium' | 'large';
+
+type Props = {
+  selectedId: string;
+  options: Option[];
+  onClick?: React.MouseEventHandler<HTMLButtonElement>;
+  size?: Size;
+};
+
+const BLOCK_NAME = 'spui-SegmentedControl';
+
+const useWindowWidth = () => {
+  const [width, setWidth] = useState(0);
+  useLayoutEffect(() => {
+    const updateWidth = () => {
+      setWidth(window.innerWidth);
+    };
+
+    window.addEventListener('resize', updateWidth);
+    updateWidth();
+
+    return () => window.removeEventListener('resize', updateWidth);
+  }, []);
+  return width;
+};
+
+export const SegmentedControl: React.FC<Props> = ({
+  selectedId: userSelectedId,
+  options,
+  onClick,
+  size = 'medium',
+}) => {
+  const [selectedId, setSelectedId] = useState(userSelectedId);
+  const [isFirstRender, setIsFirstRender] = useState(true);
+  const indicatorRef = useRef<HTMLDivElement>(null);
+  const buttonsRef = useRef<RefObject<HTMLButtonElement>[]>([]);
+  const width = useWindowWidth();
+
+  options.forEach((_, index) => {
+    buttonsRef.current[index] = createRef<HTMLButtonElement>();
+  });
+
+  useEffect(() => {
+    if (isFirstRender) {
+      setIsFirstRender(false);
+    }
+  }, [isFirstRender]);
+
+  useEffect(() => {
+    // selectedIdがどの項目にも一致しない場合は最初の項目を選択する
+    if (options.find((option) => option.id === userSelectedId) === undefined) {
+      setSelectedId(options[0].id);
+    }
+  }, [options, userSelectedId]);
+
+  useEffect(() => {
+    // 選択されたとき、あるいはブラウザのwidthが変化したときに、選択レイヤーのwidthとheight、positionを更新する
+    const indicator = indicatorRef.current;
+    const selectedButtonRef = buttonsRef.current.find(
+      (button) =>
+        button &&
+        button.current &&
+        button.current.getAttribute('aria-checked') === 'true',
+    );
+    const selectedButton = selectedButtonRef?.current;
+
+    if (indicator && selectedButton) {
+      const { offsetWidth, offsetHeight, offsetLeft } = selectedButton;
+      indicator.style.setProperty('width', `${offsetWidth}px`);
+      indicator.style.setProperty('height', `${offsetHeight}px`);
+      indicator.style.setProperty('transform', `translateX(${offsetLeft}px)`);
+    }
+  }, [selectedId, width]);
+
+  const handleClick = useCallback(
+    (e: React.MouseEvent<HTMLButtonElement>, id: string) => {
+      setSelectedId(id);
+
+      if (typeof onClick === 'function') {
+        onClick({
+          ...e,
+          currentTarget: { ...e.currentTarget, value: id },
+        });
```

Comment:
event丸ごと渡してしまえば利用者側でゴニョゴニョできるので👇で良いかなと思ったのだけどどうだろう？ 👀 
```suggestion
        onClick(e);
```

Reply:
q
eventを丸ごと渡してしまうと、labelの情報は利用者に伝わりますが、idは伝わらないと思って、95-97で `event.currentTarget.value`にidを入れています。
（前提として、labelは一意とは限らないので利用者はidを受け取る必要があります）

ただ、この点については私もまだ悩んでいて、 `event.currentTarget.value`に入れる形で果たして良いのだろうかと思っています。
この点で言えば、Design Docsにあった `(id: string) => void` 形式（もしくは、 `(event: React.MouseEvent<HTMLButtonElement>, id: string) => void` ）でうまく解決できそうだなと思っています。

Reply:
> 「labelは一意とは限らない」というのは、下記のような状況（= 同じラベルのものがる）をコンポーネントとして許容しているということですか...？

許容しない想定です。が、ロジック側で制限していないので、重複することもあるという認識です。

> また、idも（一意であることを想定してはいるものの）実装側で特に制限してはいなさそうなので、一意とは限らないと思うのですがいかがでしょうか...？ 👀

たしかにidも一意とは限らないですが、idという名称なのでユーザ側で一意にしておくべきだと思っています。


idは必ず一意であるべき（利用者が担保）で、labelは一意が望ましいが重複することもある、と考えてロジックを作っていました。
また、labelが一意であるとするのであれば、idとlabelは分ける必要もないと思いました。

Reply:
> その上で、利用者がonClickを渡したいケースって例えばどんなのが想定できますか？👀

ユーザ側のstateを更新するためにonClickが必要になると思っています（自分の想定ではほとんどのケースでonClickが必要になると認識していました）

```
const [selectedId, setSelectedId] = useState("all");

return (
<>
  <SegmentedControl 
      selectedId={selectedId}
      options={options}
      onClick={(event) => setSelectedId(event.currentTarget.value)}
   />
   {selectedId === 'all' ? <>All Content</>}
   {selectedId === 'follow' ? <>Follow Content</>}
</>
)
```

Reply:
https://github.com/openameba/spindle/pull/731#discussion_r1241379128
なるほどありがとう！
要は、「idもlabelも一意であることを想定してはいるけど、ロジック側で制限しているわけじゃない」 = 「利用者が正しい使い方で利用してくれる前提」という認識であっていますか？ 👀

labelは渡されたものを表示してるだけなので重複したものが渡されたとしてもバグったりはしませんが、
idは一意である前提のロジックなので、重複したものが渡された場合に下記のように押せなくなってしまいます 👀 

一応確認ですがこのような場合も「利用者が正しい使い方で利用してくれなかったため」として、特にコンポーネント側では担保しませんか？ 👀

https://github.com/openameba/spindle/assets/42470015/322aea8d-3be3-4ffb-98c6-963a8a8f57b6

Reply:
https://github.com/openameba/spindle/pull/731#discussion_r1241386956
こちらもありがとう！
確かにSegmentedControlの役割的に「ページ内で機能やモードを切り替える際に利用」なので、onClickはほぼ必須値みたいな形になりそうだね...！ 🙏🏻

> ただ、この点については私もまだ悩んでいて、 event.currentTarget.valueに入れる形で果たして良いのだろうかと思っています。
この点で言えば、Design Docsにあった (id: string) => void 形式（もしくは、 (event: React.MouseEvent<HTMLButtonElement>, id: string) => void ）でうまく解決できそうだなと思っています。

eventはそのまま返してあげて、event + selectedIdの両方を返す形が良いかなと思いました 🙌🏻 （Pagination同等の実装）

Reply:
> 要は、「idもlabelも一意であることを想定してはいるけど、ロジック側で制限しているわけじゃない」 = 「利用者が正しい使い方で利用してくれる前提」という認識であっていますか？ 👀

あっています

> labelは渡されたものを表示してるだけなので重複したものが渡されたとしてもバグったりはしませんが、
idは一意である前提のロジックになっていそうなので、重複したものが渡された場合に下記のように押せなくなってしまいます 👀
一応確認ですがこのような場合も「利用者が正しい使い方で利用してくれなかったため」として、特にコンポーネント側では担保しませんか？ 👀

この問題については把握していましたが、コンポーネント側の責務ではないと私は考えています。そのため何も対処しない方針です。

Reply:
`(event: React.MouseEvent, id: string) => void`で実装し直しました。tsxファイル、storybook、およびDesign Docs共に修正しました：[9addc29](https://github.com/openameba/spindle/pull/731/commits/9addc29f5eddb87d2e6be934fdb0327ee8e1d26d)

Reply:
テストの修正が漏れていたので修正しました：[d83745c](https://github.com/openameba/spindle/pull/731/commits/d83745c8ad9a194a61892d743fecedff3b424438)

Reply:
諸々考慮済みとのこと、了解です！修正もありがとうございます 🙌🏻